build_system.sh

```diff
@@ -38,70 +36,96 @@ if [ ! -f $UBUNTU_FILE ]; then
   curl -C - -o $UBUNTU_FILE $UBUNTU_BASE_URL/$UBUNTU_FILE --silent
 fi
 
-if [ "$ARCH" != "arm64" ] && [ "$ARCH" != "aarch64" ]; then
-  # Register qemu multiarch
-  docker run --rm --privileged multiarch/qemu-user-static:register --reset
+# Register qemu multiarch
+if [ "$ARCH" = "x86_64" ]; then
+  echo "Registering qemu-user-static"
+  docker run --rm --privileged multiarch/qemu-user-static:register --reset > /dev/null
 fi
 
-# Start docker build
-echo "Building image"
-export DOCKER_CLI_EXPERIMENTAL=enabled
+# Start agnos-builder docker build and create container
+echo "Building agnos-builder docker image"
 docker build -f Dockerfile.agnos -t agnos-builder $DIR
+echo "Creating agnos-builder container"
+CONTAINER_ID=$(docker container create --entrypoint /bin/bash agnos-builder:latest)
+
+# Setup mount container for macOS and CI support (namespace.so)
+echo "Building agnos-mount docker image"
+docker build -f Dockerfile.sparsify -t agnos-mount $DIR
+echo "Starting agnos-mount container"
+MOUNT_CONTAINER_ID=$(docker run -d --privileged -v $DIR:$DIR agnos-mount)
+
+# Cleanup containers on possible exit
+trap "echo \"Cleaning up containers:\"; \
+docker container rm -f $CONTAINER_ID $MOUNT_CONTAINER_ID" EXIT
 
-# Setup mount container
-MOUNT_CONTAINER_ID=$(docker run -d --privileged --volume $BUILD_DIR:$BUILD_DIR ubuntu:latest sleep infinity)
+# Define functions for docker execution
+exec_as_user() {
+  docker exec -u $USERNAME $MOUNT_CONTAINER_ID "$@"
+}
+
+exec_as_root() {
+  docker exec $MOUNT_CONTAINER_ID "$@"
+}
+
+# Create host user in container (fixes namespace.so error)
+USERNAME=$(whoami)
+exec_as_root useradd --uid $(id -u) -U -m $USERNAME &> /dev/null
```

Comment:
This breaks when the host user has uid 1000 which already exists in the docker.

Reply:
good catch! since ubuntu 24.04 images have the ubuntu 1000:1000 user. I tested this with the kernel but not system, that’s it’s always good to have a checklist. thanks!

Reply:
I have a fix and am current testing on both ubuntu and macos. PR should be up soon.

Reply:
@adeebshihadeh this needs to be merged https://github.com/commaai/agnos-builder/pull/252